Feature/session as cache

trasher opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #3763

cedric-anne commented on inc/includes.php:
```diff
@@ -105,15 +98,14 @@
    $LOADED_PLUGINS   = [];
    $plugin           = new Plugin();
-   if (!isset($_SESSION["glpi_plugins"])) {
+   if ($plugin->getPlugins() === null) {
```
`if (count($plugin->getPlugins()) === 0) {`

trasher replied:
Hum... We're originally trying to knw here if "something" has been loaded; we definitely cannot rely on the `getPlugins` method.

cedric-anne commented on inc/migration.class.php:
```diff
@@ -699,7 +699,7 @@ function executeMigration() {
 
       // as some tables may have be renamed, unset session matching between tables and classes
-      unset($_SESSION['glpi_table_of']);
+      $GLPI_CACHE->removeItem('table_of');
```
`global $GLPI_CACHE` missing.

trasher replied:
fixed

cedric-anne commented on inc/plugin.class.php:
```diff
@@ -1876,4 +1868,83 @@ static function getState($state = 0) {
+    *
+    * @return void
+    */
+   public static function setUnloadedName($name) {
```
I would have name it `setUnloadedByName`.

cedric-anne commented on inc/plugin.class.php:
```diff
@@ -347,14 +348,10 @@ function checkStates() {
    static function isAllPluginsCSRFCompliant() {
```
This method is not used (only use of it is commented).

trasher replied:
Indeed, even if it's out of the current PR scope

trasher replied:
https://github.com/glpi-project/glpi/pull/4460

cedric-anne commented on inc/plugin.class.php:
```diff
@@ -1876,4 +1868,83 @@ static function getState($state = 0) {
+      global $GLPI_CACHE;
+      $plugins = $GLPI_CACHE->getItem('plugins');
+      unset($plugins[$id]);
+      $GLPI_CACHE->replcaeItem('plugins', $plugins);
```
`replcaeItem` -> `replaceItem`

cedric-anne commented on inc/commondbtm.class.php:
```diff
@@ -126,18 +144,40 @@ static function getTable($classname = null) {
+      $classname = get_called_class();
+      if (!isset($fkeys[$classname]) || empty($fkeys[$classname])) {
+         $fkeys[$classname] = getForeignKeyFieldForTable(static::getTable());
+         $GLPI_CACHE->replaceItem('foreign_key_field_of', $fkeys);
```
Missing `global $GLPI_CACHE;`.

cedric-anne commented on inc/config.class.php:
```diff
@@ -2922,12 +2924,41 @@ public static function getCache($optname, $context = 'core') {
          }
       }
-      Toolbox::logDebug("CACHE $optname", $cache_class);
+      //Toolbox::logDebug("CACHE $optname", $cache_class);
```
If you remove this line, `$cache_class` can be removed too.

cedric-anne commented on scripts/cliupdate.php:
```diff
@@ -57,6 +57,7 @@
 include_once (GLPI_ROOT . "/inc/db.function.php");
 include_once (GLPI_CONFIG_DIR . "/config_db.php");
 
+global $PHPLOGGER, $SQLLOGGER; //globals...
```
I do not see why it is declared here.

trasher replied:
They must exists before they can be used/assigned; declaring a global at the beginning rather than in any further line does not sounds illogical to me.

trasher commented on inc/plugin.class.php:
```diff
@@ -1876,4 +1868,83 @@ static function getState($state = 0) {
+   /**
+    * Get plugins list
+    *
+    * @since 9.4
```
It is not important, and all other entries says "9.4" right now (even the one in the database). Out of scope.

cedric-anne commented on inc/includes.php:
```diff
@@ -40,25 +40,18 @@
 // Standard includes
 include_once (GLPI_ROOT . "/inc/config.php");
 
+$GLPI_CACHE->removeItems(['table_of', 'foreign_key_field_of']);
```
If items are erased on each process, maybe forcing the `Memory` adapter is a better solution.

trasher replied:
Do not remember but that looks like a dev only line :D

cedric-anne replied:
No, looks like it was already done before your modifications.

cedric-anne replied:
`glpi_foreign_key_field_of` is computed using `CommonDBTM::getTable()` function, that can be overrided by items. Maybe it has been made to be sure to do not have cache problem when updating `CommonDBTM::getTable()` results.

cedric-anne replied:
My bad, it was not cleaning but it was defining an empty array if variables was not set.

trasher commented:
>  Reading and writing it from too many places increase risk of having something that "breaks" stored data.

Cache entries are managed from a few specific methods; I do not understand "too many places".

orthagh commented:
One functional issue:
![image](https://user-images.githubusercontent.com/418844/47290562-605f4300-d600-11e8-9fe3-659e7bc5f86c.png)

On opcache & apcu, it's already not very clear, but in case of file, it uselless i think

orthagh commented on inc/commondbtm.class.php:
```diff
@@ -179,11 +194,15 @@ static function getTable($classname = null) {
+      if (!isset($glpi_tables[$classname]) || empty($glpi_tables[$classname])) {
+         $glpi_tables[$classname] = getTableForItemType($classname);
+         if ($GLPI_CACHE != null) {
+            $GLPI_CACHE->set('table_of', $glpi_tables);
```
One thing here for this entire block, as it is repeated in other place, could it be better to place the `getTableForItemType` and the cache set in the `getTablesOf` function, no ?

trasher replied:
Where is this code repeated?

I do not understand what you have in mind, to me that's not really possible. Will talk about this wit you directly.

trasher commented:
I've reworked a bit. A low memory usage is a GLPI issue, not a user/config one; so I've dropped this part. Also, high was at 50%, I've set to 80.

For the UI part...
APCu:
![screenshot_2018-10-23 glpi - setup 2](https://user-images.githubusercontent.com/224733/47340550-e92cbc00-d69e-11e8-9349-208721dffb66.png)
Filesystem:
![screenshot_2018-10-23 glpi - setup 1](https://user-images.githubusercontent.com/224733/47340556-f0ec6080-d69e-11e8-8b23-0c2e630f6c62.png)
Memory:
![screenshot_2018-10-23 glpi - setup](https://user-images.githubusercontent.com/224733/47340568-f77ad800-d69e-11e8-97e0-ae0cb4aa6e0c.png)

Cannot test other adapters.

orthagh commented:
perfect for ui
